pkg/api/v1beta3/types.go

```diff
@@ -934,7 +934,7 @@ type ServiceSpec struct {
 	Ports []ServicePort `json:"ports" description:"ports exposed by the service"`
 
 	// This service will route traffic to pods having labels matching this selector. If null, no endpoints will be automatically created. If empty, all pods will be selected.
-	Selector map[string]string `json:"selector" description:"label keys and values that must match in order to receive traffic for this service; if empty, all pods are selected, if not specified, endpoints must be manually specified"`
+	Selector string `json:"selector" description:"label keys and values that must match in order to receive traffic for this service; if empty, all pods are selected, if not specified, endpoints must be manually specified"`
```

Comment:
@smarterclayton  my understanding is that a map has three limitations: 
1. Only one operator per `<mykey,value>` pair. Today this is `=` or `in` or `==`. 
2. In a map we cannot have multiple requirements `<mykey,value1>`,  `<mykey,value2>` for the same key, since a map model a 1:1... well is a map.
3. A map has an implicit operator to `join` all the requirements. Today this operator is the AND.

So, due to limitation 1 we cannot store in API: `mykey notin (value)` since the operator `notin` is not represented. See https://github.com/GoogleCloudPlatform/kubernetes/issues/3627
Due to limitation 2 we cannot store in API a query like `mykey,mykey notin (value)` (i.e. `mykey` exists and must not be equal to `value`),  which cannot be represented anyway (due to 1) but it's another kind of limitation: 2 requirements with the same key: `mykey`

The array of `labels.Requirements` solves the two first problems. It introduces all the operators defined in [labels package](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/labels/selector.go#L53)  and remove the limitation of the uniqueness of the key. Hence  we'll be able to store `mykey, mykey notin (yourvalue)`.
The implicit operator (3) issue remains and due to this we cannot write a selector like: `mykey=value1 or mykey=value2`.
To write `OR` operator we  need to extend the selector query language adding `or` or `not` or parenthesis.

About the `all`: I would introduce it since we already have in the code `labels.Everything()` for this.
In that case an empty string may represent the endpoints to be manually specified and the non empty string should represent the `labels.Selector` if correctly parsed. But I'm not sure we'll cover all the needs. Am I missing something?

@thockin 
thanks I don't find so ugly :) it's a way to handle two representation structured and unstructured. I'm OK with this.
I keep to think that  representing the selector as a string is the only way to reduce the compatibility issue (and discussion like this :) ) if early or late a decision is taken to modify the selector language. 

A possible example of powerful (maybe too rich) query language in REST service could be [this](https://confluence.atlassian.com/display/JIRA/Advanced+Searching#AdvancedSearching-AND)